4.1.1rc

jngrad opened the pull request:
4.1.1 bugfix release.

jngrad committed:
Update Doxygen and Sphinx docs for quaternions

kosovan committed:
Merge #3184

3184: some modifications to constant-pH tutorial

Co-authored-by: Peter Košovan <peter.kosovan@natur.cuni.cz>
Co-authored-by: Jonas Landsgesell <jonaslandsgesell@users.noreply.github.com>
Co-authored-by: Peter Košovan <kosovan@icp.uni-stuttgart.de>
Co-authored-by: Jean-Noël Grad <jgrad@icp.uni-stuttgart.de>
Co-authored-by: Kai Szuttor <kai@icp.uni-stuttgart.de>

KaiSzuttor committed:
pylint:fixed: Dangerous default value [] as argument.

KaiSzuttor committed:
pylint: fixed: redefinition of function.

jngrad committed:
Fix broken LB thermostat checkpointing mechanism

jngrad committed:
Fix test logic

junghans committed:
cmake: install missing targets

junghans committed:
cmake: remove SOVERSION

junghans committed:
cmake: fix permission of python module install

junghans committed:
cmake: generate version.hpp at compile time to avoid rebuilds

mkuron committed:
Fix mpiio with stdlibc++ range checking

mkuron committed:
Formatting

mkuron committed:
use iterator arithmetic inside mpiio_common_read

reinaual committed:
added test for second charged system tutorial

reinaual committed:
fixed typo

mkuron committed:
ESS2019 installation guide updates

mkuron committed:
More macOS installation documentation

mkuron committed:
CMake syntax fix

fweik committed:
maintainer: Escape module python in wrapper script

mkuron committed:
rst title levels

mkuron committed:
Remove pip from install instructions

KaiSzuttor committed:
doc: fix sections.

reinaual committed:
added mean value check for charge test

reinaual committed:
changed test criteria

reinaual committed:
updated docstring

reinaual committed:
formatting

jngrad committed:
Fix links in release notes

jngrad committed:
Remove duplicate argument documentation

Argument descriptions already in the docstrings of the linked
functions.

jngrad committed:
Fix typos

jngrad committed:
Provide links to Dockerfiles for other Linux OSes

Based on feedback from the ESPResSo summer school's user and
developer meeting (@kosovan).

reinaual committed:
removed useless mean density check

reinaual committed:
changed LJ to WCA in nacl tutorial script

reinaual committed:
replaced LJ with WCA in charged_system-1 tutorial

reinaual committed:
added proper equil.time to tutorial

KaiSzuttor committed:
Update CMakeLists.txt

jngrad committed:
Fix issue in the NPT documentation

Fix broken link, incorrect docstrings and copy-paste errors (based
on inspection of the C++ code for the NPT integrator and offline
discussion with @christophlohrmann).

jngrad committed:
Refactor checkpoint test logic

jngrad committed:
Add getter for integrator state

jngrad committed:
Fix checkpointing mechanism for integrators

Note: in Python, a function taking positional arguments can take
named arguments in any order.

jngrad committed:
Fix checkpointing mechanism for NPT thermostat

jngrad committed:
Add checkpoint test for integrators + thermostats

jngrad committed:
Remove unused integrator + thermostat code logic

jngrad committed:
Add feature guards in tests

jngrad committed:
Skip UBSAN error (division by zero)

Happens when calling `system.integrator.run(0)`. In that case the
average instantaneous pressure is undefined and can only be NaN.

jngrad committed:
Document CMake logic

jngrad committed:
Fix version nomenclature in Sphinx docs

jngrad committed:
Deprecate DESTDIR install method

Installing ESPResSo with `make install DESTDIR="/path/to/es"` will
install files in the wrong folder and set an incorrect runtime path
in Cython shared objects, causing import errors. See full report:
https://github.com/espressomd/espresso/pull/3228#issuecomment-541422452

jngrad committed:
Fix cluster_analysis module

When installing ESPResSo, the C++ shared objects and Cython shared
objects are stored in the same folder. Since they have the same .so
extension, when importing espressomd.cluster_analysis, the file
cluster_analysis.py is loaded and the file cluster_analysis.so is
automatically considered a Cython shared object, and Python fails
to find the Cython-specific symbol PyInit_cluster_analysis. This
is remediated by giving a different name to the C++ shared object.

jngrad committed:
Add job to check the installed version of ESPResSo

Install the files in a different directory and check the python and
tutorials/samples tests can still run with the installed files.
Please note that pypresso will experience a major slowdown in CI
when located outside the build directory, hence the 3h timeout.
This job is meant to be run only during releases and in PRs that
add new core modules.

jngrad committed:
Skip tutorial ferrofluid part 3 (timeout)

Takes more than 1 hour to run in CI when pypresso is located
outside the build directory.

jngrad committed:
Typos

jngrad committed:
Remove redundant PYTHONPATH

jngrad committed:
Simplify CMake test logic

The installed files only exist if the Python bindings are enabled.

jngrad committed:
Remove unused DESTDIR variable

jngrad committed:
Cannot test python3 in CI

The clang:6.0 job fails due to undefined ASAN symbols.
The logic to fix that error is in the pypresso wrapper.

jngrad committed:
Merge #3228 into 4.1.1rc

3228: Fix cmake install

Description of changes:
- fix the CMake build system to include libraries in the list of installed files and test it in a dedicated CI job
- fix the versioning of Cython .so shared objects
- simplify the CMake tests

jngrad committed:
Refactor NPT interface

Allow bool values in the script interface. The default value for
box dimension fluctuations is now [True, True, True] instead of
the counter-intuitive [0, 0, 0] that was implicitly converted to
[1, 1, 1] in the core.

jngrad committed:
Update integrator documentation

jngrad committed:
Make NpT cubic_box a boolean flag

jngrad committed:
Remove NpT parameter p_inst_av

Average instantaneous pressure: instantaenous pressure divided by
the number of steps in the integration (NaN if nsteps=0). Unused in
the core and observables, inaccessible from the Python interface.

junghans commented on testsuite/cmake/test_install.sh:
```diff
@@ -16,21 +16,24 @@
   done
+
+  # check the espressomd module can be imported from pypresso
+  assert_return_code "@CMAKE_INSTALL_PREFIX@/bin/pypresso" -c "import espressomd"
```
```suggestion
  assert_return_code "@CMAKE_INSTALL_BINDIR@/pypresso" -c "import espressomd"
```

jngrad replied:
Done.

junghans commented on src/python/pypresso.cmakein:
```diff
@@ -8,9 +8,9 @@
 else
-  PYTHONPATH=@PYTHON_DIR@
+  PYTHONPATH="@PYTHON_DIR@"
 fi
```
Just use
```
PYTHONPATH="@PYTHON_DIR@${PYTHONPATH:+:}$PYTHONPATH"
```
instead.

jngrad replied:
I don't know how portable this variable expansion syntax is, only tried it on a few shells. It's also quite cryptic; we had an issue once where a long-standing Travis-CI error message was finally solved by removing an incorrect use of that syntax (#2932). I think the current verbose if/else version is better from a maintainability perspective.

junghans replied:
Cryptic, but posix.

junghans replied:
In #2932, you used `${var+x}`, not `${var:+x}`, which makes a huge different, as it doesn't check for empty but defined variables.

Looking at #2932, you really need to run the bashism script on the scripts, that don't explicit use `bash` (or change them all to `bash`) and in the `bash` script, I would always use `[[  ]]` instead of `[ ]` as it automatically quotes the left hand side. And using brace-quotes don't replace double quotes unfortunately.

jngrad replied:
I had plans to reduce bashisms (#3242), but there is more interest in our team to actually move away from shell scripts entirely ([ES meeting Oct 16](https://github.com/espressomd/espresso/wiki/Espresso-meeting-2019-10-16#code-quality-and-linting)).

junghans replied:
You guys should enable codacy, that helped us a lot to incrementally improve the shell scripts.

jngrad replied:
I did not spot the difference between `${var+x}` and `${var:+x}` until you pointed it out to me. This is an issue inherent to shell. I don't mind using advanced variable substitution tricks in my personal shell scripts, but it's a different story on collaborative project. We aim to lower the entry barrier for new contributors on espresso, and this is an example where a simple if/else is more explicit than a not well-known variable substitution, in my opinion.

We used to have automatic code linting and checking with `pylint` and `CodeFactor` in the past, but they were eventually removed from CI a year ago. This is something I would like to experiment with again in 4.2.

junghans replied:
https://app.codacy.com/manual/junghans/espresso/dashboard

junghans commented on testsuite/cmake/CMakeLists.txt:
```diff
@@ -12,29 +12,16 @@ function(CMAKE_TEST)
-  add_custom_target(setup_install_prefix COMMAND make install WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
-  add_dependencies(setup_install setup_install_prefix)
-  cmake_test(FILE test_install_prefix.sh DEPENDENCIES BashUnitTests.sh)
+  cmake_test(FILE test_install.sh DEPENDENCIES BashUnitTests.sh)
```
It would be useful to check if any `.py` files got installed outside of `@PYTHON_DIR@`.

jngrad replied:
Do you mean if a `.py` file was installed by mistake somewhere in `/tmp/espresso-unit-tests` outside the `espressomd` folder?

junghans replied:
Exactly!

jngrad replied:
Done.

jngrad committed:
Don't decay NpT integrator to NVT integrator

When the NpT integrator is initialized with incorrect values, halt
script execution with an exception instead of initializing NVT.

jngrad committed:
Cleanup comments

jngrad committed:
Fix regression

RudolfWeeber commented:
>I had plans to reduce bashisms ( <https://github.com/espressomd/espresso/issues/3242> #3242), but there is more interest in our team to actually move away from shell scripts entirely ( <https://github.com/espressomd/espresso/wiki/Espresso-meeting-2019-10-16#code-quality-and-linting> ES meeting Oct 16).

There remains the question of relative priority compared to other tasks, though.

Unless someone commits to rewriting the scripts in Python quite soon, existing work on improving them could be merged, I think.

jngrad committed:
Document parsing PDB with MDAnalysis

jngrad committed:
Load standard GNU variables in CMake

jngrad committed:
Add extra CMake install path check

jngrad committed:
Use GNU variable BINDIR

jngrad committed:
Remove obsolete Python guard

The Python guard isn't necessary anymore because submodules are
now installed in the Python package espressomd, and the testsuite
itself already has a Python guard in the top-level CMakeLists.txt.

jngrad committed:
Use modern CMake install procedure

Remove the non-standard CMAKE_INSTALL_EXEC_PREFIX variable in
favor of CMAKE_INSTALL_PREFIX. Install espressomd in the standard
python3.X folder instead of the deprecated python3 folder. Remove
duplicated code logic. Use explicit keyword arguments in calls to
functions from distutils.sysconfig.

jngrad committed:
Get branch/commit/state as string, not bytestring

jngrad committed:
Narrow include of version.hpp

junghans commented:
Please, let me know once you tagged the 4.1.1rc and I will give it a test build on Fedora Rawhide.

junghans committed:
fix PROJECT_VERSION expansion

jngrad committed:
Cleanup CMake

Fix indentation, rename output variables, remove superfluous quotes
around string variables. CMake quotes string variables implicitly
when passed to functions, even if they contain whitespaces. Quotes
are only necessary when setting a string variable.

jngrad committed:
Cleanup docs and fix typos

jngrad committed:
Fix velocity Verlet interface

When setting up the NPT integrator, then the VV integrator, the core
integrator remains on NPT. Since NVT and VV are the same, we can set
the core integrator to NVT when the interface is set to VV.

fweik committed:
utils: Re-add uniform test + docstring

jngrad committed:
Typos

jngrad committed:
Fix overshadowed test

RudolfWeeber committed:
LB Boundary force: fix unit conversion and tests

Reference solutions for forces on shear planes and for stokes drag need dynamic viscosity, but LB uses kinetmatic viscosity.
This switches the tests to dynamic viscosity and removes a 1/rho from the unit conversion in the script interface.
Also, a force balance test is added, which does not depend on the density.

RudolfWeeber committed:
Remove leaked-in debug code

fweik committed:
core: Simplify particle removal

christophlohrmann committed:
fix writevtk() bug where types=all was ignored

christophlohrmann committed:
Apply suggestions from code review

switch conditionals as suggested by JNGrad

Co-Authored-By: Jean-Noël Grad <jgrad@icp.uni-stuttgart.de>

hirschsn committed:
Use && instead of test -a

POSIX does not specify test for > 4 arguments:
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/test.html

hirschsn committed:
Avoid re-splitting of arguments

fweik commented:
LGTM

jngrad committed:
Partial Merge #3277

3277: Removed ENGINE shear torque calculation

Description of the partial merge:
 - Removed rotational_friction
 - Fixed engine_lb test
 - Adaped docs

Co-authored-by: Florian Weik <fweik@icp.uni-stuttgart.de>

jngrad committed:
Merge #2748

2748: Additional LB test cases

* Buyancy force on a sphere in a box of fluid under gravity
* Momentum conservation in a fluid with an external volume force applied and a particle with an oposite force applied

Co-authored-by: Rudolf Weeber <weeber@icp.uni-stuttgart.de>
Co-authored-by: Florian Weik <fweik@icp.uni-stuttgart.de>

jngrad committed:
Update copyright year

jngrad committed:
Spellcheck and grammar check

jngrad committed:
Add 4.1.1 release notes

jngrad commented:
@junghans: merged, tagged and released. The official announcement will me made in a few hours.
https://github.com/espressomd/espresso/releases/tag/4.1.1

junghans commented:
@jngrad thanks, testing on Fedora: https://koji.fedoraproject.org/koji/taskinfo?taskID=38967783
